networkit/cpp/graph/ParallelGraphBuilder.cpp

```diff
@@ -0,0 +1,107 @@
+
+#include <omp.h>
+#include <stdexcept>
+
+#include <networkit/graph/ParallelGraphBuilder.hpp>
+
+namespace NetworKit {
+
+ParallelGraphBuilder::ParallelGraphBuilder(count n, bool directed)
+    : n(n), directed(directed), outEdgesPerThread(omp_get_max_threads()),
+      inEdgesPerThread(directed ? omp_get_max_threads() : 0) {
+    for (int i = 0; i < omp_get_max_threads(); i++) {
+        outEdgesPerThread[i].resize(omp_get_max_threads());
+    }
+    if (directed) {
+        for (int i = 0; i < omp_get_max_threads(); i++) {
+            inEdgesPerThread[i].resize(omp_get_max_threads());
+        }
+    }
+}
+
+void ParallelGraphBuilder::addEdgeParallel(index a, index b) {
+    if (directed) {
+        addHalfOutEdgeParallel(Unsafe{}, a, b);
+        addHalfInEdgeParallel(Unsafe{}, b, a);
+    } else {
+        addHalfEdgeParallel(Unsafe{}, a, b);
+        addHalfEdgeParallel(Unsafe{}, b, a);
+    }
+}
+
+void ParallelGraphBuilder::addHalfEdgeParallel(Unsafe, index a, index b) {
+    assert(a != b);
+    outEdgesPerThread[omp_get_thread_num()][a % omp_get_max_threads()].emplace_back(HalfEdge{a, b});
```

Comment:
If you use emplace this should be enough, same for the other functions below:
```suggestion
    outEdgesPerThread[omp_get_thread_num()][a % omp_get_max_threads()].emplace_back(a, b);
```

Reply:
the current implementation of ´HalfEdge` does not have such a constuctor

Reply:
Good idea to add it then. Improves readability.